test/unit/test_test_await_methods.gd

```diff
@@ -80,79 +87,86 @@ class TestTheNewWaitMethods:
+		var result = await wait_for_signal(signaler.the_signal, 10)
+		assert_true(result)
+
+	func test_wait_for_signal_returns_false_when_signal_not_emitted():
 		var signaler = add_child_autoqfree(TimedSignaler.new())
-		signaler.the_signal.connect(node.queue_free)
-		signaler.emit_after(.1)
-
-		await assert_eventually(is_freed, 1.0)
-
-		assert_between(counter.time, 0.0, .2)
-		assert_freed(node)
-
-# ------------------------------------
-# Could not get these to trigger the error I was trying to replicate.  This was
-# a useful exercies though and I'm not ready to part with this code.  So it is
-# here as an example of using parameterized tests to control the order of the
-# execution of tests with different behavior per iteration.
-# ------------------------------------
-# class TestYieldTimerResetAfterSignalEmitted:
-# 	extends GutTest
-
-# 	var _wait_data = ParameterFactory.named_parameters(
-# 		['kind', 'time'],
-# 		[
-# 			['s', 3.0],
-# 			['t', 4.0],
-# 		])
-
-# 	var _counter = null
-# 	func before_each():
-# 		_counter = add_child_autoqfree(Counter.new())
-
-
-# 	func test_one(param=use_parameters(_wait_data)):
-# 		print('time left = ', gut._yield_timer.time_left)
-# 		var signaler = add_child_autoqfree(TimedSignaler.new())
-# 		var t = param.time
-
-# 		if(param.kind == 's'):
-# 			signaler.emit_after(t)
-# 			await wait_for_signal(signaler.the_signal, t + 2)
-# 		else:
-# 			await wait_seconds(t)
-
-# 		assert_between(_counter.time, t - 0.1, t + 0.1)
-# 		print('time left = ', gut._yield_timer.time_left)
\ No newline at end of file
+		signaler.emit_after(10)
+		var result = await wait_for_signal(signaler.the_signal, .5)
+		assert_false(result)
+
+
+	func test_wait_until_waits_ends_when_method_returns_true():
+		var all_is_good = func():
+			return counter.time > .25
+
+		await wait_until(all_is_good, .5)
+		assert_almost_eq(counter.time, .25, .05)
+		assert_false(did_wait_timeout(), 'did_wait_timeout')
+
+	func test_wait_until_times_out():
+		var all_is_good = func():
+			return false
+
+		await wait_until(all_is_good, .5)
+		assert_almost_eq(counter.time, .5, .05)
+		assert_true(did_wait_timeout(), 'did_wait_timeout')
+
+	func test_wait_until_returns_true_when_it_finishes():
+		var all_is_good = func():
+			return counter.time > .25
+
+		var result = await wait_until(all_is_good, .5)
+		assert_true(result)
+
+	func test_wait_until_returns_false_when_it_times_out():
+		var all_is_good = func():
+			return false
+
+		var result = await wait_until(all_is_good, .5)
+		assert_false(result)
+
+	func test_wait_until_accepts_string_as_thrid_arg():
+		var pred_methods = PredicateMethods.new()
+		var method = pred_methods.called_x_times.bind(10)
+
+		await wait_until(method, 1.1, 'DID YOU SEE THIS?')
+		pass_test("Check output for DID YOU SEE THIS?")
+
+	func test_wait_until_accepts_time_between():
+		var pred_methods = PredicateMethods.new()
+		var method = pred_methods.called_x_times.bind(10)
+
+		await wait_until(method, 1.1, .25)
```

Comment:
Why have the 3rd param be either msg_or_time_between?
At first glance I found it a bit confusing.

I wish GDScript had [named parameters](https://github.com/godotengine/godot/issues/6866#issue-183827265) for stuff like this
Maybe pass in a Dict parameter for all the "[options](https://docs.godotengine.org/en/stable/classes/class_dictionary.html)" (timeout_s, time_between_s, message, etc)
```
wait_until({"timeout_s": 1.1, "time_between_s": .25, "message": "bleerg"})
```
Then any option could be omitted and use defaults values.

Not a big deal, just food for though

Reply:
Named parameters would be great.  There's a couple places in GUT where I've had to take this approach and I've never been happy with it.  It's always ugly trying to pick which should be the last optional parameter and what they should be named.  Maybe just `p3` and `p4` would be less confusing?  They don't tell you anything but they aren't misleading either.

Honestly, the time-between feature smells of feature creep.  I talked myself into it because it seems more difficult to implement later.  And if I did add it later, the method signature would end up looking this way so it was backwards compatible (a horrible justification, I know).

I've thought about dictionaries for optional parameters.  I think they are a bad user experience though, they are annoying to type and more vulnerable to typos.  On the implementation side you have to validate the dictionary before you do anything.  GUT can cause a test to fail with an error, but you don't want the end user to have to check return values just for valid parameters.  It's all stuff that can be worked around, but its never felt worth it for a few rare cases.

Maybe it should just be two methods:  `wait_until`, `wait_until_with_time_between` (it's such a long name though).

Naming things is the worst.